crates/librqbit/src/blocklist.rs

```diff
@@ -0,0 +1,314 @@
+use tokio::{io::AsyncBufReadExt, io::BufReader};
+use tokio_util::io::StreamReader;
+use tracing::{debug, info, trace};
+
+pub struct Blocklist {
+    // ipv4 and ipv6 do not overlap
+    // see: https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5
+    blocked_ranges: IntervalTree<IpAddr, ()>,
+}
+
+impl Blocklist {
+    pub fn empty() -> Self {
+        return Self::new(std::iter::empty());
+    }
+
+    pub fn new(ip_ranges: impl IntoIterator<Item = std::ops::Range<IpAddr>>) -> Self {
+        Self {
+            blocked_ranges: IntervalTree::from_iter(ip_ranges.into_iter().map(|r| (r, ()))),
+        }
+    }
+
+    pub async fn load_from_url(url: &str) -> Result<Self> {
+        let response = reqwest::get(url).await.map_err(|e| anyhow::anyhow!(e))?;
+        if response.status() != 200 {
+            return Err(anyhow::anyhow!(
+                "Failed to fetch blocklist: HTTP {}",
+                response.status()
+            ));
+        }
+
+        let content_length = response
+            .content_length()
+            .ok_or_else(|| anyhow::anyhow!("Failed to get content length"))?;
+
+        if content_length < 2 {
+            return Err(anyhow::anyhow!(
+                "Content too short: not enough data to determine compression"
+            ));
+        }
+
+        let reader = StreamReader::new(
+            response
+                .bytes_stream()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
+        );
+        Self::create_from_stream(reader).await
+    }
+
+    pub async fn load_from_file(path: &str) -> Result<Self> {
+        let file = tokio::fs::File::open(path).await?;
+        let reader = tokio::io::BufReader::new(file);
+        Self::create_from_stream(reader).await
+    }
+
+    async fn create_from_stream<R>(reader: R) -> Result<Self>
+    where
+        R: AsyncRead + Unpin + Send,
+    {
+        let mut peek_bytes = [0u8; 2];
+        let mut reader = tokio::io::BufReader::new(reader);
+
+        // Peek the first bytes by filling buffer
+        let buffer = reader.fill_buf().await?;
+        if buffer.len() >= 2 {
+            peek_bytes.copy_from_slice(&buffer[0..2]);
+        } else {
+            return Err(anyhow::anyhow!(
+                "Content too short: not enough data to determine compression"
+            ));
+        }
+
+        // Check for Gzip magic bytes (1F 8B)
+        let is_gzip = peek_bytes == [0x1F, 0x8B];
+
+        let reader: Pin<Box<dyn AsyncRead + Send>> = if is_gzip {
+            trace!("Detected Gzip file, decompressing...");
+            Box::pin(BufReader::new(GzipDecoder::new(reader)))
+        } else {
+            trace!("Plain text file detected.");
+            Box::pin(reader)
+        };
+
+        let reader = BufReader::new(reader);
+        let mut lines = reader.lines();
+        let mut ip_ranges: Vec<std::ops::Range<IpAddr>> = Vec::new();
+        while let Some(line) = lines.next_line().await? {
```

Comment:
lines() allocates every line. Instead you could use read_line() from AsyncBufReadExt and reuse one allocation